tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,231 @@
+// <copyright file="AsyncHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Datadog.Trace.Debugger.Instrumentation;
+
+namespace Datadog.Trace.Debugger.Helpers
+{
+    internal static class AsyncHelper
+    {
+        private const string StateMachineNamePrefix = "<";
+        private const string StateMachineNameSuffix = ">";
+        private const string StateMachineFieldsPrefix = "<>";
+        private const string BuilderFieldName = StateMachineFieldsPrefix + "t__builder";
+        private const string StateFieldName = StateMachineFieldsPrefix + "1__state";
+        private const string StateMachineThisFieldName = StateMachineFieldsPrefix + "4__this";
+        private const string StateMachineLiveDebuggerAddedField = "<>dd_liveDebugger_isReEntryToMoveNext";
+        private const BindingFlags AllFieldsBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldNameValueType[] GetHoistedArgumentsFromStateMachine<TTarget>(TTarget instance, ParameterInfo[] originalMethodParameters)
+        {
+            if (originalMethodParameters is null || originalMethodParameters.Length == 0)
+            {
+                return Array.Empty<FieldNameValueType>();
+            }
+
+            var foundFields = new FieldNameValueType[originalMethodParameters.Length];
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+
+            int found = 0;
+            for (int i = 0; i < allFields.Length; i++)
+            {
+                // we have more fields than parameters, so we break once we found all parameters
+                if (found == foundFields.Length)
+                {
+                    break;
+                }
+
+                var field = allFields[i];
+                // early exit here based on generated name pattern https://datadoghq.atlassian.net/browse/DEBUG-928
+                if (field.Name.StartsWith(StateMachineNamePrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                for (var j = 0; j < originalMethodParameters.Length; j++)
+                {
+                    if (field.Name == originalMethodParameters[j].Name)
+                    {
+                        foundFields[j] = new FieldNameValueType(field.Name, field.GetValue(instance), field.FieldType);
```

Comment:
What about generic params? `field.GetValue` might break on all kind of scenarios.